src/libraries/System.Net.Http/src/System/Net/Http/Headers/HeaderDescriptor.cs

```diff
@@ -15,32 +16,67 @@ namespace System.Net.Http.Headers
     // Use HeaderDescriptor.TryGet to resolve an arbitrary header name to a HeaderDescriptor.
     internal readonly struct HeaderDescriptor : IEquatable<HeaderDescriptor>
     {
-        private readonly string _headerName;
-        private readonly KnownHeader? _knownHeader;
+        /// <summary>
+        /// Either a <see cref="KnownHeader"/> or <see cref="string"/>
+        /// </summary>
+        private readonly object _descriptor;
 
         public HeaderDescriptor(KnownHeader knownHeader)
         {
-            _knownHeader = knownHeader;
-            _headerName = knownHeader.Name;
+            _descriptor = knownHeader;
         }
 
         // This should not be used directly; use static TryGet below
         internal HeaderDescriptor(string headerName)
         {
-            _headerName = headerName;
-            _knownHeader = null;
+            _descriptor = headerName;
         }
 
-        public string Name => _headerName;
-        public HttpHeaderParser? Parser => _knownHeader?.Parser;
-        public HttpHeaderType HeaderType => _knownHeader == null ? HttpHeaderType.Custom : _knownHeader.HeaderType;
-        public KnownHeader? KnownHeader => _knownHeader;
+        public string Name
+        {
+            get
+            {
+                object? descriptor = _descriptor;
+                return descriptor is KnownHeader knownHeader ?
+                    knownHeader.Name :
+                    Unsafe.As<string>(descriptor);
+            }
+        }
+
+        public object Descriptor => _descriptor;
+
+        public HttpHeaderParser? Parser => _descriptor is KnownHeader knownHeader ? knownHeader.Parser : null;
+        public HttpHeaderType HeaderType => _descriptor is KnownHeader knownHeader ? knownHeader.HeaderType : HttpHeaderType.Custom;
+        public string Separator => _descriptor is KnownHeader knownHeader ? knownHeader.Separator : HttpHeaderParser.DefaultSeparator;
+
+        public bool Equals(HeaderDescriptor other)
+        {
+            object? descriptor = _descriptor;
+            object? otherDescriptor = other._descriptor;
```

Comment:
Why `object?`

Reply:
I'm not sure what you had in mind? `_descriptor` is an `object?`.

Reply:
It does not look like that https://github.com/dotnet/runtime/pull/62981/files#diff-67d75c27e0a1d09ec8892c8e8f5aa5ce532a38d53090fb7e422a40530605a252R22

Reply:
Right, but `HeaderDescriptor` is a struct, so if you access `Name` on a default instance, you may as well get `null`.
I'll change the field to `object?` since we are using default instances of it elsewhere.